File: 588_design_in_memory_file_system.java
class FileSystem {
    private static class Node {
        private final String name;
        private final Map<String, Node> nodeByName = new TreeMap<>(); // if node is directory
        private final StringBuilder content = new StringBuilder(); // if node if file
        private boolean isFile;

        Node(final String name){
            this.name = name;
        }
    }

    private final Node root = new Node("");

    public FileSystem() {}
    
    public List<String> ls(String path) {
        Node node = traversePath(path);
        return node.isFile ? List.of(node.name) : List.copyOf(node.nodeByName.keySet());
    }
    
    public void mkdir(String path) {
        traversePath(path);
    }
    
    public void addContentToFile(String filePath, String content) {
        Node file = traversePath(filePath);
        file.isFile = true;
        file.content.append(content);
    }
    
    public String readContentFromFile(String filePath) {
        return traversePath(filePath).content.toString();
    }

    private Node traversePath(String path){
        Node current = root;
        String[] pathLst = path.split("/");
        
        for(int i = 1; i<pathLst.length; i++){
            final int ic = i;
            current = current.nodeByName.computeIfAbsent(pathLst[i], k -> new Node(pathLst[ic]));
        }
        return current;
    }
}

/**
 * Your FileSystem object will be instantiated and called as such:
 * FileSystem obj = new FileSystem();
 * List<String> param_1 = obj.ls(path);
 * obj.mkdir(path);
 * obj.addContentToFile(filePath,content);
 * String param_4 = obj.readContentFromFile(filePath);
 */

File: 588_design_in_memory_file_system.py
class Node:
    def __init__(self, name, is_file = False):
        self.name = name
        self.children_map = {}
        self.is_file = is_file
        self.file_content = ''

class FileSystem:
    def __init__(self):
        self.root_node = Node('')

    def ls(self, path: str) -> List[str]:
        fileNode = self._traverse(path)
        if fileNode.is_file:
            return [fileNode.name]
        else:
            return sorted(fileNode.children_map.keys())

    def mkdir(self, path: str) -> None:
        self._traverse(path)

    def addContentToFile(self, filePath: str, content: str) -> None:
        fileNode = self._traverse(filePath)
        fileNode.is_file = True
        fileNode.file_content += content

    def readContentFromFile(self, filePath: str) -> str:
        fileNode = self._traverse(filePath)

        return fileNode.file_content
    
    # fetch node of the filePath. If doesn't exist, create path
    def _traverse(self, filePath: str) -> Node:
        if filePath == '/': # root
            return self.root_node
        
        path = filePath.split('/')
        node = self.root_node
        for p in path[1:]:
            if p not in node.children_map:
                node.children_map[p] = Node(p)
            node = node.children_map[p]
        return node

# Your FileSystem object will be instantiated and called as such:
# obj = FileSystem()
# param_1 = obj.ls(path)
# obj.mkdir(path)
# obj.addContentToFile(filePath,content)
# param_4 = obj.readContentFromFile(filePath)
